megatron/training/training.py

```diff
@@ -1324,33 +1324,63 @@ def evaluate_and_print_results(prefix, forward_step_func,
+        collected_non_loss_data=dict()
+        for name in data_iterator.keys():
+            total_loss_dict[name], collected_non_loss_data[name], timelimit = evaluate(
+                forward_step_func, data_iterator[name], model,
+                process_non_loss_data_func, config, verbose)
+        
+        string=''
+        for name in data_iterator.keys():
+            string += 'dataset ' + name + ' | '
+            string += 'validation loss at {} | '.format(prefix)
+            for key in total_loss_dict[name]:
+                string += '{} value: {:.6E} | '.format(key, total_loss_dict[name][key].item())
+                ppl = math.exp(min(20, total_loss_dict[name][key].item()))
+                string += '{} PPL: {:.6E} | '.format(key, ppl)
+                if writer and is_last_rank():
+                    writer.add_scalar(f'validation/{key}/{name}',
+                                    total_loss_dict[name][key].item(),
+                                    iteration)
+                    writer.add_scalar(f'validation vs samples/{key}/{name}',
+                                    total_loss_dict[name][key].item(),
+                                    args.consumed_train_samples)
+                    if args.log_validation_ppl_to_tensorboard:
+                        writer.add_scalar(f'validation ppl/{key}/{name} ', ppl,
+                                        iteration)
+                        writer.add_scalar(f'validation ppl vs samples/{key}/{name}',
+                                        ppl, args.consumed_train_samples)
+            string+='\n'
+    else:
+        total_loss_dict, collected_non_loss_data, timelimit = evaluate(
+            forward_step_func, data_iterator, model,
+            process_non_loss_data_func, config, verbose)
+        # Timelimit hit during evaluation
+        if timelimit:
+            return
+        string = ' validation loss at {} | '.format(prefix)
+        for key in total_loss_dict:
+            string += '{} value: {:.6E} | '.format(key, total_loss_dict[key].item())
+            ppl = math.exp(min(20, total_loss_dict[key].item()))
+            string += '{} PPL: {:.6E} | '.format(key, ppl)
+            if writer:
+                writer.add_scalar('{} validation'.format(key),
+                                total_loss_dict[key].item(),
+                                iteration)
+                writer.add_scalar('{} validation vs samples'.format(key),
+                                total_loss_dict[key].item(),
+                                args.consumed_train_samples)
+                if args.log_validation_ppl_to_tensorboard:
+                    writer.add_scalar('{} validation ppl'.format(key), ppl,
+                                    iteration)
+                    writer.add_scalar('{} validation ppl vs samples'.format(key),
+                                    ppl, args.consumed_train_samples)
+                if wandb_writer and is_last_rank():
```

Comment:
Dunno if worth it, but if we add this to the the case with multiple_valid_sets, we can potentially move from tensorboard to wandb

Reply:
I prefer tensorboard but I added the option. should be working